File: generated-code/codecraft/cpp/TcpReadWrite/model/EntityProperties.cpp
#include "EntityProperties.hpp"

EntityProperties::EntityProperties(int size, int buildScore, int destroyScore, bool canMove, int populationProvide, int populationUse, int maxHealth, int initialCost, int sightRange, int resourcePerHealth, std::optional<BuildProperties> build, std::optional<AttackProperties> attack, std::optional<RepairProperties> repair) : size(size), buildScore(buildScore), destroyScore(destroyScore), canMove(canMove), populationProvide(populationProvide), populationUse(populationUse), maxHealth(maxHealth), initialCost(initialCost), sightRange(sightRange), resourcePerHealth(resourcePerHealth), build(build), attack(attack), repair(repair) { }

// Read EntityProperties from input stream
EntityProperties EntityProperties::readFrom(InputStream& stream) {
    int size = stream.readInt();
    int buildScore = stream.readInt();
    int destroyScore = stream.readInt();
    bool canMove = stream.readBool();
    int populationProvide = stream.readInt();
    int populationUse = stream.readInt();
    int maxHealth = stream.readInt();
    int initialCost = stream.readInt();
    int sightRange = stream.readInt();
    int resourcePerHealth = stream.readInt();
    std::optional<BuildProperties> build = std::optional<BuildProperties>();
    if (stream.readBool()) {
        build = BuildProperties::readFrom(stream);
    }
    std::optional<AttackProperties> attack = std::optional<AttackProperties>();
    if (stream.readBool()) {
        attack = AttackProperties::readFrom(stream);
    }
    std::optional<RepairProperties> repair = std::optional<RepairProperties>();
    if (stream.readBool()) {
        repair = RepairProperties::readFrom(stream);
    }
    return EntityProperties(size, buildScore, destroyScore, canMove, populationProvide, populationUse, maxHealth, initialCost, sightRange, resourcePerHealth, build, attack, repair);
}

// Write EntityProperties to output stream
void EntityProperties::writeTo(OutputStream& stream) const {
    stream.write(size);
    stream.write(buildScore);
    stream.write(destroyScore);
    stream.write(canMove);
    stream.write(populationProvide);
    stream.write(populationUse);
    stream.write(maxHealth);
    stream.write(initialCost);
    stream.write(sightRange);
    stream.write(resourcePerHealth);
    if (build) {
        stream.write(true);
        const BuildProperties& buildValue = *build;
        buildValue.writeTo(stream);
    } else {
        stream.write(false);
    }
    if (attack) {
        stream.write(true);
        const AttackProperties& attackValue = *attack;
        attackValue.writeTo(stream);
    } else {
        stream.write(false);
    }
    if (repair) {
        stream.write(true);
        const RepairProperties& repairValue = *repair;
        repairValue.writeTo(stream);
    } else {
        stream.write(false);
    }
}

// Get string representation of EntityProperties
std::string EntityProperties::toString() const {
    std::stringstream ss;
    ss << "EntityProperties { ";
    ss << "size: ";
    ss << size;
    ss << ", ";
    ss << "buildScore: ";
    ss << buildScore;
    ss << ", ";
    ss << "destroyScore: ";
    ss << destroyScore;
    ss << ", ";
    ss << "canMove: ";
    ss << canMove;
    ss << ", ";
    ss << "populationProvide: ";
    ss << populationProvide;
    ss << ", ";
    ss << "populationUse: ";
    ss << populationUse;
    ss << ", ";
    ss << "maxHealth: ";
    ss << maxHealth;
    ss << ", ";
    ss << "initialCost: ";
    ss << initialCost;
    ss << ", ";
    ss << "sightRange: ";
    ss << sightRange;
    ss << ", ";
    ss << "resourcePerHealth: ";
    ss << resourcePerHealth;
    ss << ", ";
    ss << "build: ";
    if (build) {
        const BuildProperties& buildValue = *build;
        ss << buildValue.toString();
    } else {
        ss << "none";
    }
    ss << ", ";
    ss << "attack: ";
    if (attack) {
        const AttackProperties& attackValue = *attack;
        ss << attackValue.toString();
    } else {
        ss << "none";
    }
    ss << ", ";
    ss << "repair: ";
    if (repair) {
        const RepairProperties& repairValue = *repair;
        ss << repairValue.toString();
    } else {
        ss << "none";
    }
    ss << " }";
    return ss.str();
}

File: generated-code/rust/codecraft/TcpReadWrite/src/model/entity_properties.rs
use super::*;

/// Entity properties
#[derive(Clone, Debug)]
pub struct EntityProperties {
    /// Size. Entity has a form of a square with side of this length
    pub size: i32,
    /// Score for building this entity
    pub build_score: i32,
    /// Score for destroying this entity
    pub destroy_score: i32,
    /// Whether this entity can move
    pub can_move: bool,
    /// Number of population points this entity provides, if active
    pub population_provide: i32,
    /// Number of population points this entity uses
    pub population_use: i32,
    /// Maximum health points
    pub max_health: i32,
    /// Cost to build this first entity of this type. If this is a unit (entity can move), the cost is increased by 1 for each existing unit of this type
    pub initial_cost: i32,
    /// If fog of war is enabled, maximum distance at which other entities are considered visible
    pub sight_range: i32,
    /// Amount of resource added to enemy able to collect resource on dealing damage for 1 health point
    pub resource_per_health: i32,
    /// Build properties, if entity can build
    pub build: Option<model::BuildProperties>,
    /// Attack properties, if entity can attack
    pub attack: Option<model::AttackProperties>,
    /// Repair properties, if entity can repair
    pub repair: Option<model::RepairProperties>,
}

impl trans::Trans for EntityProperties {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.size.write_to(writer)?;
        self.build_score.write_to(writer)?;
        self.destroy_score.write_to(writer)?;
        self.can_move.write_to(writer)?;
        self.population_provide.write_to(writer)?;
        self.population_use.write_to(writer)?;
        self.max_health.write_to(writer)?;
        self.initial_cost.write_to(writer)?;
        self.sight_range.write_to(writer)?;
        self.resource_per_health.write_to(writer)?;
        self.build.write_to(writer)?;
        self.attack.write_to(writer)?;
        self.repair.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let size: i32 = trans::Trans::read_from(reader)?;
        let build_score: i32 = trans::Trans::read_from(reader)?;
        let destroy_score: i32 = trans::Trans::read_from(reader)?;
        let can_move: bool = trans::Trans::read_from(reader)?;
        let population_provide: i32 = trans::Trans::read_from(reader)?;
        let population_use: i32 = trans::Trans::read_from(reader)?;
        let max_health: i32 = trans::Trans::read_from(reader)?;
        let initial_cost: i32 = trans::Trans::read_from(reader)?;
        let sight_range: i32 = trans::Trans::read_from(reader)?;
        let resource_per_health: i32 = trans::Trans::read_from(reader)?;
        let build: Option<model::BuildProperties> = trans::Trans::read_from(reader)?;
        let attack: Option<model::AttackProperties> = trans::Trans::read_from(reader)?;
        let repair: Option<model::RepairProperties> = trans::Trans::read_from(reader)?;
        Ok(Self {
            size,
            build_score,
            destroy_score,
            can_move,
            population_provide,
            population_use,
            max_health,
            initial_cost,
            sight_range,
            resource_per_health,
            build,
            attack,
            repair,
        })
    }
}